afqinsight/nn/pt_models.py

```diff
@@ -0,0 +1,401 @@
+            nn.Linear(input_shape, 500),
+            nn.ReLU(),
+            nn.Dropout(0.2),
+            nn.Linear(500, 500),
+            nn.ReLU(),
+            nn.Dropout(0.2),
+            nn.Linear(500, 500),
+            nn.ReLU(),
+            nn.Dropout(0.3),
+            nn.Linear(500, n_classes)
+        )
+        
+        if output_activation == torch.softmax:
+            self.output_activation = nn.Softmax(dim=1)
+        else:
+            self.output_activation = None
+        
+        if verbose:
+            print(self.model)
+
+    def forward(self, x):
+        x = self.model(x)
+        if self.output_activation:
+            x = self.output_activation(x)
+        return x
+    
+def MLP4(input_shape, n_classes):
+    mlp4 = mlp4(input_shape, n_classes, output_activation = torch.softmax, verbose=False)
+    return mlp4
+    
+class cnn_lenet(nn.Module):
+    def __init__(self, input_shape, n_classes, output_activation = torch.softmax, verbose=False):
+        self.n_conv_layers = int(round(math.log(input_shape[0], 2)) - 3)
+        if verbose:
+            print(f"Pooling layers: {self.n_conv_layers}")
+        
+        #idea is to create a list that sequential can use to
+            #create the model with the order of layers?
+        conv_layers = []
+
+        for i in range(self.n_conv_layers): 
+            if i == 0:
+                conv_layers.append(
+                    nn.Conv1d(
+                        # what is in channels?
+                        in_channels= input_shape[1],
+                        out_channels=6,
+                        kernel_size=3,
+                        padding=1, # because padding in tensorflow is 'same' and strides = 1
+                    )
+                )
+            else:
+                conv_layers.append(
+                    nn.Conv1d(
+                        # what is in channels?
+                        in_channels= 6 + 10 * i,
+                        out_channels=6 + 10 * i,
+                        kernel_size=3,
+                        padding=1, # because padding in tensorflow is 'same' and strides = 1
+                    )
+                )
+            conv_layers.append(nn.ReLU())
+            conv_layers.append(nn.MaxPool1d(kernel_size=2))
+            
+        self.model = nn.Sequential(
+            *conv_layers,
+            nn.Flatten(),
+            nn.Linear(input_shape, 120),
+            nn.ReLU(),
+            nn.Dropout(0.5),
+            nn.Linear(120, 84),
+            nn.ReLU(),
+            nn.Dropout(0.5),
+            nn.Linear(84, n_classes),
+        )
+        
+        if output_activation == torch.softmax:
+            self.output_activation = nn.Softmax(dim=1)
+        else:
+            self.output_activation = None
+        
+
+    def forward(self, x):
+        x = self.model(x)
+        if self.output_activation:
+            x = self.output_activation(x)
+        return x
+
+    def cnn_lenet(input_shape, n_classes):
```

Comment:
Maybe this bit of boilerplate -- for each class, you implement a function that initializes an instance of this class -- could be automated? Have a function that takes class names as inputs and generates these functions as its outputs?